### ISSUE TITLE
Undesirable throughput/scaling characteristics of Orchestration

### ISSUE THREAD

**Issue by olitomlinson** [Azure/azure-functions-durable-extension#779]
@cgillum I could really do with your opinion on this.

**TLDR**: At any point in time, I'm observing that the current Orchestrations throughput* performance is directly correlated to the current throughput of a Service Bus Trigger that is _starting_ said Orchestrations.

*throughput = speed of scheduling activities and receiving response from activities*

---

I have a Durable Function that has been in production for a couple of months now and as time passes I'm observing some throughput trends that I need to address and resolve.

I've included a sequence diagram of the Orchestration at the bottom of this post, if you wish to understand how this orchestration works, but It is quite simple.

I will present 3 graphs that highlight this issue.

I have instrumented my Orchestration code so I can observe the timings for significant operations :

* Just before and just after `await starter.StartNewAsync()` (triggering the orchestration)
* Just before and just after `await context.CallActivityWithRetryAsync()` (inside the orchestration)
* Just before and just after `await context.WaitForExternalEventAsync()` (inside the orchestration)

----

**Graph 1 - approx 6 weeks ago**

![image](https://user-images.githubusercontent.com/4224880/58477066-8d517680-814a-11e9-93e9-e850d946744b.png)

This graph represents a  continuous ramp-up of orchestrations starting. The graph is **not** cumulative, each data point on the Y-axis is exclusive.

Ignore the blue line above, this represents receiving of events from an external system.

Notice above how all other data points are closely linked almost hidden by the green line? This means that the time spent between scheduling an activity and the activity being executed and control returning to the orchestration is almost instant.

To see it more clearly, below, I strip away all the telemetry generated _inside_ the orchestration, and leave just telemetry that represents _just before_ and _just after_ `await starter.StartNewAsync()`

**Compare this graph below with the one above - notice how they correlate almost exactly. It would appear that orchestration throughput is maintaining pace, despite the ramp-up in orchestrations being started - perfect!**

![image](https://user-images.githubusercontent.com/4224880/58477615-dce47200-814b-11e9-81bc-dfc7fb4c44bb.png)

----

**Graph 2 - Present**

This is where things start getting interesting...

We've optimised our upstream services at this point, which means we can have more 'work' ready to complete sooner. The red line below represents Orchestrations _just before_ and _just after_ `await starter.StartNewAsync()`. its a much more aggressive ramp-up of work, with more orchestrations starting sooner.

![image](https://user-images.githubusercontent.com/4224880/58479012-46b24b00-814f-11e9-9aba-de440f3f5131.png)

Now lets highlight the first activity call in the pink line below

![image](https://user-images.githubusercontent.com/4224880/58479370-14551d80-8150-11e9-8a3a-7e7375408f83.png)

To reiterate, the pink line is _just before_ a call to `await context.CallActivityWithRetryAsync()`

Notice above how the pink line (just before the first activity call) is wildly different from the red line (just before orchestration start). This is not typical with the 1st graph from 6 weeks ago. I am sympathetic to the fact that are more orchestrations being started sooner, so we might be hitting some upper limit with the internals of DF.

**But more interestingly, notice how the Pink line dives around `22:10`. It dives at _exactly_ the same time as the Service Bus Trigger work starts to drop (Service Bus queue length is now 0)**

**Why should the throughput of the pink line dive? Why is it not maintaining the pace it had prior to the dive?**

Lets look at the telemetry from _just before_  **second** activity call below (blue line)

![image](https://user-images.githubusercontent.com/4224880/58480015-5a5eb100-8151-11e9-9e7a-16aa69fce08c.png)

It has the same drop in throughput at `22:10`

Notice how the tail of events is now getting low and long. Initially I thought that this might have been to do with my Activities taking a while to complete. But my telemetry tells me that those Activities are completing in around 17-30ms so I am confident that this is not something External to the Durable Functions environment that is causing the slow down, it appears to be internal. Is work sat on the internal DF queues and not being processed in time, possibly causing the slow down?

----

**Graph 3 - Present - Testing a theory**

My theory at this point is that when the Service Bus workload depletes towards 0, the scale controller starts removing instances of my Durable Function App, thereby reducing the overall throughput of the control-queues.

I tested this theory, by starting 18k orchestrations, waiting 10 minutes, then starting another 18k.

The red line tracks _just before_ and _just after_ calls being made to `starter.StartNewAsync()`

The first red hump is the first batch of 18,000 orchestrations being started. Notice how the throughput tracks that.

The second red hump is the second batch of 18,000 orchestrations being started. Notice once again how the throughput tracks that.

Once the Service Bus queue is empty, at approx `07:25` throughput gradually decreases into the long and low tail once again.

![image](https://user-images.githubusercontent.com/4224880/58480666-b544d800-8152-11e9-9423-007ac3c9b64d.png)

----

**My conclusion**

Overall throughput of in-flight durable function orchestrations is affected by the Service Bus trigger queue length/ processing rate. Likely to be the scale controller adding more hosts, to cope with the Service Bus queue length, which in turn, is boosting the throughput of the in-flight orchestrations.

**The problem here is that once the service bus queue is empty, the throughput drops for the in-flight orchestrations and this is not desirable at all for my use-case. In fact, I won't be able to use Durable Functions for this use-case if this is the intended scaling behavior, which will be very sad indeed.**

**Why isn't the scale controller not recognising the fact that the `control-queue` s are growing, and forcing the addition of more hosts?** 

It feels like the scale of the Function App has been overly-linked / overly-biased to the initial Service Bus queue that fed it the work load.

---- 

**Orchestration Sequence Diagram**

![image](https://user-images.githubusercontent.com/4224880/58476917-20d67780-814a-11e9-9019-03e520b7519b.png)

----

**Other info**
UK south 
DF 1.8.0
Azure Functions ~2
.NET c#

**Comment by olitomlinson** [comment]
I've drilled into my App Insights telemetry to observe the Functions Runtime version and my findings can be summarised with the following timeline.

* Prior to, and around, 24th of April
**Good throughput**
Runtime [v2.0.12408](https://github.com/Azure/azure-functions-host/releases/tag/v2.0.12408)

* Around 1st of May
**Inconclusive throughput** - not enough samples
Runtime [v2.0.12427](https://github.com/Azure/azure-functions-host/releases/tag/v2.0.12427)

* Around 5th of May, onwards
**Poor throughput**
Runtime [v2.0.12438](https://github.com/Azure/azure-functions-host/releases/tag/v2.0.12438)

**Comment by cgillum** [comment]
Very interesting. I'm curious to know if anything else besides just the Functions runtime changed - for example, if there was an App Service platform update which corresponds to this. Would it be possible for you to privately send me the name of your function app so I can investigate further (perhaps a DM on Twitter? I really wish GitHub made this easier...).

Your theory about what's going on is interesting, but the scale controller should not scale you down unless it feels confident that you are over-scaled. In the case of the **control queue**, you should never need more VMs than the number of partitions (only one VM can read from a control-queue at a time), so I don't expect scaling-down due to Service Bus load completing should impact your ability to handle Durable workloads. It could possibly impact the latency of scheduling and running activity functions (i.e. the **work-item queue**), but from your description, it sounds like that's not the problem.

Do you have any data that shows specific VM counts over time?  If not, I can try to dig up that info.

**Comment by olitomlinson** [comment]
DMd.

So its impossible for multiple partitions to be leased to a single Host? Is it always 1 partition to 1 Host?

I don't have anything to hand regards VM counts, not entirely sure how I'll derive that, but I'll take a look in the morning to see if I can find anything.

If you want to see the code in detail or any of the telemetry let me know. 

If it helps, the telemetry I provided in **graph 3**  was in the following UTC time range

```
| where timestamp > todatetime('2019-05-28T07:00:00.000')
| where timestamp < todatetime('2019-05-28T08:00:00.000')
```

**Comment by cgillum** [comment]
It is possible for multiple partitions to be leased by a single host. That happens, for example, when you only have a single host in the consumption plan - it listens to all queues. However, you'll never have a case where multiple hosts lease a single partition. For that reason, the scale controller never scales out a function app beyond the number of partitions, because those extra hosts would just sit idle (unless there were lots of items in the work-item queue to process - but there is separate scaling logic to handle that).

**Comment by olitomlinson** [comment]
I just can’t grok what I’m seeing from the graphs. Occams Razor suggests my graphs are plain wrong but I’ve had them checked by one of my Seniors...

From graph 3, We can clearly see that there is available capacity, otherwise that second  hump would have yielded nothing. And why such a vast difference from the instantaneous performance of graph 1?

Couple of other thoughts floating in my head...

1. When starting a new orchestration, does that intent go straight onto the designated control-queue for the given instanceId, or does the intent get added to the work-queue, which is later resolved to the correct control-queue?

2. Can a host that is leasing one or more control-queue also be simultaneously receiving work-queue work? Or are the two mutually exclusive by design? If a host can do both simultaneously, is it being accidentally scaled in when the work-queue tends towards 0, therefore temporarily removing control-queue leases/capacity at the same tine?

3. When an activity is in the process of completing, does the framework signal back direct to the designated control-queue? I have a feeling it does, just want to check with you.

**Comment by olitomlinson** [comment]
@cgillum I've reverted my Durable Function back to 2.0.12408 (my last known good version) and re-ran the exact same scenario from **Graph 3** and it looks like the performance is back to a good state as it was in **Graph 1**.

![2 0 12408-insights](https://user-images.githubusercontent.com/4224880/58544101-8d0fb480-81f8-11e9-9cd4-74651d216ed4.PNG)

I'm going to run a few more tests on various Runtime versions to build certainty that this is caused by a regression in the Runtime.

**Comment by olitomlinson** [comment]
### v2.0.12427

Once again, good throughput can be observed as all the metrics are tightly packed together. 

There is no 'second hump' due to the time it took for ServiceBusTrigger to scale out and receive the work as explained in the second graph below.

![image](https://user-images.githubusercontent.com/4224880/58549536-a66a2e00-8203-11e9-88e8-ed9f171dfb98.png)

It took the Function app a little more time to scale-out here, as work was sitting on the Service Bus queue for longer, not as desirable as the 2.0.12408 results 

![image](https://user-images.githubusercontent.com/4224880/58549184-dc5ae280-8202-11e9-98ec-95aa420bd03e.png)

**Comment by olitomlinson** [comment]
### v2.0.12438

I had an expectation that this version would revert back to the bad throughput that we have experienced recently, but it hasn't, it looks good. Although the data points are a little less compact around `10:47` - `10:48`

Is the Runtime version a red herring?

![image](https://user-images.githubusercontent.com/4224880/58552068-660dae80-8209-11e9-9dfc-2ca8e2f1216f.png)

**Comment by olitomlinson** [comment]
### v2.0.12493

Upgraded to the most recent runtime version and ran the same test, with the same results.

![image](https://user-images.githubusercontent.com/4224880/58565193-35883d80-8226-11e9-8aba-06b785a31962.png)

**Comment by olitomlinson** [comment]
Given the testing across all instances, It looks like its not related to :

1. Azure Functions Runtime Version (Tested old and new)
2. Application Code (unchanged)
3. Environment (unchanged)

The only thing that still remains a question is from my original point in **Graph 2** of the original post, which correlated a drop in the service bus message processing with the throughput of the orchestration function.

Given that orchestration throughput is now keeping up with the rate of orchestration starts (for some unexplained reason) I can't recreate the original issues provided by **Graph 2** and **Graph 3**

**Comment by cgillum** [comment]
Responding to some of the questions you asked earlier:

> 1. When starting a new orchestration, does that intent go straight onto the designated control-queue for the given instanceId, or does the intent get added to the work-queue, which is later resolved to the correct control-queue?

When starting a new orchestration, the intent goes straight onto the designated control queue. The work-item queue is not involved.

> 2. Can a host that is leasing one or more control-queue also be simultaneously receiving work-queue work? Or are the two mutually exclusive by design? If a host can do both simultaneously, is it being accidentally scaled in when the work-queue tends towards 0, therefore temporarily removing control-queue leases/capacity at the same tine?

Yes, all hosts can receive messages from the work-item queue, even if it has leased on or more control queues. However, in order for the scale controller to scale-in the number of hosts, _both_ the work-item and control queues must be trending towards 0. If the work-item queue is trending towards zero and the control queue is steady, then the scale controller will not scale things in.

> 3. When an activity is in the process of completing, does the framework signal back direct to the designated control-queue? I have a feeling it does, just want to check with you.

Yes, the result of an activity function execution is a message that gets directly added to the designated control-queue.

**Comment by olitomlinson** [comment]
@cgillum I've decided to make steps towards testing my theory that these issues are linked to the throughput of the Service Bus Trigger that is starting the orchestrations.

I started pulling apart my function app into two discrete function Apps, that both referenced the same TaskHub. 

One app contains the Service bus triggers which houses the call to `client.StartNewAsync()` and another Function App that houses orchestration Function and various activities.

It looked like it was going to work until the function app that contains the service bus trigger, complained until at runtime, I get the following exception coming from the Function App that contains the call to `client.StartNewAsync()` :

`The function '<REDACTED>' doesn't exist, is disabled, or is not an orchestrator function. Additional info: No orchestrator functions are currently registered!` - which makes sense as the Orchestration Function lives in the other Function App.

So now I'm looking through the DF code and find this exception originates from [WebJobs.Extensions.DurableTask/DurableTaskExtension.cs#L617](https://github.com/Azure/azure-functions-durable-extension/blob/f2e92b866aed1d085eed33d2a7bf50d9aecfb70d/src/WebJobs.Extensions.DurableTask/DurableTaskExtension.cs#L617])

This called from [OrchestrationMiddleware Method](https://github.com/Azure/azure-functions-durable-extension/blob/f2e92b866aed1d085eed33d2a7bf50d9aecfb70d/src/WebJobs.Extensions.DurableTask/DurableTaskExtension.cs#L325) and from within `this.config.ThrowIfFunctionDoesNotExist` contained in [StartNewAsync() Method](https://github.com/Azure/azure-functions-durable-extension/blob/f2e92b866aed1d085eed33d2a7bf50d9aecfb70d/src/WebJobs.Extensions.DurableTask/DurableOrchestrationClient.cs#L81)

----
My question Chris, If I remove the call to 'this.config.ThrowIfFunctionDoesNotExist' from within `StartNewAsync`, could I trick the Extension into thinking that the orchestration does actually exist, which will then subsequently place the intent to start an orchestration on a `control` queue, which will then be picked up the by the _real_ Orchestration running in the other function App?

**Comment by cgillum** [comment]
@olitomlinson based on the code snippet you linked to, that check is there to ensure that we have a means of invoking the function that implements the orchestrator. If you comment out the check, then you would get a `NullReferenceException` [here](https://github.com/Azure/azure-functions-durable-extension/blob/f2e92b866aed1d085eed33d2a7bf50d9aecfb70d/src/WebJobs.Extensions.DurableTask/DurableTaskExtension.cs#L335).

I suspect there may actually be a problem with your setup. or you may be investigating the wrong code. If your Service Bus function app doesn't have any orchestration or activity triggers, then this code path should never get invoked.

**Comment by olitomlinson** [comment]
@cgillum  In `StartNewAsync` it checks a _KnownOrchestrators_ list, I imagine this list is populated when the Function App comes up.

As the orchestration function is no longer registered in the same Functions App as where ‘StartNewAsync’ is being called, it correctly throws the exception when calling StartNewAsync.

So my theory is that if I can comment out this check to see if the orchestration is known, then I can actually get the ‘start’ intent into a control queue and have it invoked by the _real_ orchestration function on the other App. Does this sound feasible?

**Comment by olitomlinson** [comment]
@cgillum Ok I managed to build a custom version of the Durable Function Extension so I could split apart the orchestration and activities, from the service bus trigger function. [I did this by simply removing this check here](https://github.com/Azure/azure-functions-durable-extension/blob/f2e92b866aed1d085eed33d2a7bf50d9aecfb70d/src/WebJobs.Extensions.DurableTask/DurableOrchestrationClient.cs#L81)

So my theory is wrong, the Service Bus Trigger has **no impact** on the throughput of the Durable Function. 

Although the graph below still demonstrates identical performance problems as previously reported :

![image](https://user-images.githubusercontent.com/4224880/59061573-b6d17700-889b-11e9-9585-fb3919decc44.png)

However the slow down is still visible, at least I know now its absolutely not tied to the Service Bus Trigger

----

I'll break down the graph to demonstrate what I believe is the orchestration throughput exponentially declining.

The **blue line** represents telemetry that is emitted from an Activity, just before the activity completes. 

The **green line** represents when the orchestration regains control after awaiting the activity.

![image](https://user-images.githubusercontent.com/4224880/59062133-0ebcad80-889d-11e9-97b7-2a825cce0f74.png)

Notice how the at the tail of the blue and green lines there is approx 10 minutes delay in the activity completing and the orchestration resuming. This gap appears to grow over time.

----

Here is another example. This is the **second activity**  in the orchestrations life time that begins as soon as the first activity above completes.

![image](https://user-images.githubusercontent.com/4224880/59062910-dae28780-889e-11e9-94a3-3e0f0f2db263.png)

Once again, notice how the at the tail of the blue and green lines there is approx 10 minutes delay in the activity completing and the orchestration resuming.

----

I know for a fact that none of my activities are taking this long to complete. Most of them complete within 10 - 50ms as they are doing very light work.

So whats causing the slow down?

The telemetry tells me the 'work' queue appears to be processed extremely quickly as activities start almost milliseconds from being scheduled by the orchestration. My only guess is that the `control` queues aren't gradually being processed less and less.

**Comment by olitomlinson** [comment]
Ran the same test with `"enableHttpTriggerExtendedInfoCollection": false` and now I'm getting the positive throughput I expected!

Could be a coincidence so I'm going to run the same test a few more times to gain certainty.

![image](https://user-images.githubusercontent.com/4224880/59063675-893afc80-88a0-11e9-932b-e07e4bfb35f5.png)

**Comment by olitomlinson** [comment]
And again, it looks back to normal when compared to results from 2 months ago at the top of this issue.

![image](https://user-images.githubusercontent.com/4224880/59065661-25670280-88a5-11e9-99a1-d75c08e98287.png)

**Comment by olitomlinson** [comment]
**3rd test** -- still looking great!

![image](https://user-images.githubusercontent.com/4224880/59068284-ade8a180-88ab-11e9-8885-edb46f636fa2.png)

**Comment by olitomlinson** [comment]
**4th test** - very positive again

![image](https://user-images.githubusercontent.com/4224880/59099526-394d4b80-891b-11e9-8de7-a9e69da338e4.png)

**Comment by cgillum** [comment]
Great! So do you think it's safe to say that the undesirable throughput is a result of the perf regression caused by the Application Insights change in the Azure Functions runtime? If so, do you think it makes sense for us to close out this issue?

**Comment by olitomlinson** [comment]
@cgillum yes I’m pretty sure that this is the cause. Since I disabled the setting I’ve not had observed a recurrence. I’ll close this. 

However I am in a bind right now with regards to allowing the runtime version to continue to be on ~2. I _want_ and _need_ to trust that the Functions Team has enough measures in place to prevent this kind of throughput degradation from happening again, but I don’t have that confidence right now.

I really hope they can improve the situation! If you could pass on this feedback to the team, that would be great.

I really appreciate all your efforts Chris et al :)

**Comment by brettsam** [comment]
Hey @olitomlinson -- we're still trying to track down exactly why so many people were seeing issues with this app insights setting. You seemed to be able to reproduce it pretty reliably. We'd love to try to build a repro based on what you had -- if you'd be interested, we can do it over email (my email is in my github profile). Thanks for all the digging here!

**Comment by olitomlinson** [comment]
@brettsam I can share the source with you if you like but it won't be build-able due to private packages etc Is that enough? Its going to take me a few days to get this to you so please bare with me.

**Comment by brettsam** [comment]
Yes that may be enough. I've tried several other customer's solutions and haven't yet been able to reproduce the perf issues that they're seeing. It feels like there's some piece missing that we've not yet noticed. More data would be great.

You haven't tried removing the setting and re-running your scenario above since this was closed, have you? I'm curious if it still reproduces the issue.

**Comment by olitomlinson** [comment]
@brettsam Can confirm that since removing the setting, performance is not negatively affected. Looks to be fixed 👍

**Comment by kylemccullen** [comment]
@olitomlinson I'm working on a project that is using orchestrated activities and want to use application insights to provide data to give better insight into how the app is running. Charts similar to yours would be helpful. I do have telemetry creating custom events at the start and end of activities. I'm able to query and display charts, but not quite the way I want. What queries are you running to display these charts in app insights?
